### PR TITLE
feat: Floating point exception monitoring

### PR DESCRIPTION
**Pull request by paulgessinger**

### CODE REVIEW

**Comment by paulgessinger** [comment]
I think this is ready for review now. @andiwand @noemina ?

File: Examples/Framework/src/Framework/Sequencer.cpp [hunk 1]
```diff
@@ -8,17 +8,24 @@
...
+#include <cfenv>
 #include <chrono>
 #include <exception>
+#include <limits>
 #include <numeric>
 #include <stdexcept>
 
+#include <math.h>
+#include <stdio.h>
+#include <stdlib.h>
```
Comment (andiwand) [inline Examples/Framework/src/Framework/Sequencer.cpp:27]: should we replace these c includes with c++ ones?

File: Examples/Io/Root/src/RootPropagationStepsWriter.cpp [hunk 2]
```diff
@@ -169,27 +169,27 @@ ActsExamples::ProcessCode ActsExamples::RootPropagationStepsWriter::writeT(
       double actor = step.stepSize.value(Acts::ConstrainedStep::actor);
       double aborter = step.stepSize.value(Acts::ConstrainedStep::aborter);
       double user = step.stepSize.value(Acts::ConstrainedStep::user);
-      double act2 = actor * actor;
-      double acc2 = accuracy * accuracy;
-      double abo2 = aborter * aborter;
-      double usr2 = user * user;
+      double actAbs = std::abs(actor);
```
Comment (andiwand) [inline Examples/Io/Root/src/RootPropagationStepsWriter.cpp:172]: hm i guess I missed those with my regex. might be worth adopting it

File: Examples/Scripts/Python/full_chain_odd.py [hunk 3]
```diff
@@ -41,91 +41,102 @@
 field = acts.ConstantBField(acts.Vector3(0.0, 0.0, 2.0 * u.T))
 rnd = acts.examples.RandomNumbers(seed=42)
 
-s = acts.examples.Sequencer(events=100, numThreads=-1, outputDir=str(outputDir))
+with acts.FpeMonitor():
```
Comment (andiwand) [inline Examples/Scripts/Python/full_chain_odd.py:44]: should we also do this to the ITk full chain?

File: Examples/Scripts/Python/geometry.py [hunk 4]
```diff
@@ -87,8 +88,10 @@ def runGeometry(
 
 
 if "__main__" == __name__:
-    detector, trackingGeometry, decorators = AlignedDetector.create()
+    #  detector, trackingGeometry, decorators = AlignedDetector.create()
```
Comment (andiwand) [inline Examples/Scripts/Python/geometry.py:91]:
```suggestion
    # detector, trackingGeometry, decorators = AlignedDetector.create()
```

**Review by andiwand — COMMENTED** [review]
can you ad a little summary in the description?

I dropped a few comments. otherwise should be good to go

File: thirdparty/backward-cpp/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,10 @@
+cmake_minimum_required( VERSION 3.11 )
+include( FetchContent )
+
+# Declare where to get VecMem from.
```
[the remark this answers is not in the document]
Comment (paulgessinger) [reply thirdparty/backward-cpp/CMakeLists.txt:4]: Needs to be removed again.

File: Tests/UnitTests/Core/Utilities/FpeMonitorTests.cpp [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+#include <optional>
+
+namespace utf = boost::unit_test;
+
+namespace Acts {
+
+namespace Test {
+
+// BOOST_AUTO_TEST_CASE(FpeMonitorTest) {
```
[the remark this answers is not in the document]
Comment (paulgessinger) [reply Tests/UnitTests/Core/Utilities/FpeMonitorTests.cpp:21]: I'll refactor this test.

File: CI/physmon/physmon.py [hunk 7]
```diff
@@ -62,198 +62,80 @@
...
-)
-detector, trackingGeometry, decorators = getOpenDataDetector(
-    getOpenDataDetectorDirectory(), matDeco
-)
-digiConfig = srcdir / "thirdparty/OpenDataDetector/config/odd-digi-smearing-config.json"
-geoSel = srcdir / "thirdparty/OpenDataDetector/config/odd-seeding-config.json"
-
-
-field = acts.ConstantBField(acts.Vector3(0, 0, 2 * u.T))
+with acts.FpeMonitor():
```
[the remark this answers is not in the document]
Comment (paulgessinger) [reply CI/physmon/physmon.py:65]: You mean to much indentation?

File: Core/src/Material/Interactions.cpp [hunk 8]
```diff
@@ -324,7 +325,8 @@ float Acts::computeEnergyLossLandauSigmaQOverP(const MaterialSlab& slab,
   //           = (1/p)^4 * (q/beta)² * var(E)
   // do not need to care about the sign since it is only used squared
   const auto pInv = qOverP / q;
-  return std::sqrt(rq.q2OverBeta2) * pInv * pInv * sigmaE;
+  assert(rq.q2OverBeta2 >= 0 && "Negative q2OverBeta2");
+  return clampValue<float>(std::sqrt(rq.q2OverBeta2) * pInv * pInv * sigmaE);
```
[the remark this answers is not in the document]
Comment (paulgessinger) [reply Core/src/Material/Interactions.cpp:329]: Well it throws an FPE if I don't fix it here? Or did I mess something up?

File: Examples/Framework/src/Framework/Sequencer.cpp (hunk 1 above)
Comment (paulgessinger) [reply Examples/Framework/src/Framework/Sequencer.cpp:27]: Yeah I think this is just again my vim auto-including.

File: Examples/Python/tests/root_file_hashes.txt [hunk 9]
```diff
@@ -22,66 +22,56 @@ test_itk_seeding__fatras_particles_final.root: b044d57a0e8e79673e40e362379d6fdc8
 test_itk_seeding__fatras_particles_initial.root: a294ff442940c8929bef67fb8921f21a69fe2645f0ce7828a54c2f8b4d7e1c90
 test_propagation__propagation_steps.root: 174301b25784dbb881196b658f2d7f99c0a2ea688a0129e6110fc19aa5cf8e54
 test_material_recording__geant4_material_tracks.root: a0b1c45d687041963aba359793ce7b67b8dde7f9a95cf34eb95f709bdb70ca64
-test_truth_tracking_kalman[generic-0.0]__trackstates_fitter.root: bbf815d14d1d0cec87e2e78423df40305dfe192af5fd29caedb5e57a2592b515
-test_truth_tracking_kalman[generic-0.0]__tracksummary_fitter.root: 7f222f81c121da1105ca4a1791297b7fbcdcb901fc7e742b115866551ad2a79f
+test_truth_tracking_kalman[generic-0.0]__trackstates_fitter.root: cbaf0a29893eecec7a429fb74325a961eadb160346bd88933f7e481e3b3d24b6
```
[the remark this answers is not in the document]
Comment (paulgessinger) [reply Examples/Python/tests/root_file_hashes.txt:25]: I believe so.

File: Examples/Python/tests/helpers/__init__.py [hunk 10]
```diff
@@ -47,6 +47,14 @@
...
     edm4hepEnabled = False
 
+try:
+    import acts.examples.pythia8
+
+    pythia8Enabled = True
+except ImportError:
+    pythia8Enabled = False
+
+
```
Comment (paulgessinger) [inline Examples/Python/tests/helpers/__init__.py:57]: Removing this in favor of #1739

File: Examples/Scripts/Python/full_chain_odd.py (hunk 3 above)
Comment (paulgessinger) [reply Examples/Scripts/Python/full_chain_odd.py:44]: We don't run it in the CI though. But wouldn't hurt I guess?

File: Examples/Scripts/Python/geometry.py (hunk 4 above)
Comment (paulgessinger) [reply Examples/Scripts/Python/geometry.py:91]: Reverting this entire file.

File: Core/include/Acts/Seeding/SeedFinder.ipp [hunk 11]
```diff
@@ -405,6 +405,10 @@ void SeedFinder<external_spacepoint_t, platform_t>::createSeedsForGroup(
 
         // use geometric average
         float cotThetaAvg2 = cotThetaB * cotThetaT;
+        if (cotThetaAvg2 < 0) {
```
[the remark this answers is not in the document]
Comment (paulgessinger) [reply Core/include/Acts/Seeding/SeedFinder.ipp]: Removed.

File: CI/physmon/physmon.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (andiwand) [reply CI/physmon/physmon.py]: yeah exactly

File: Examples/Scripts/Python/full_chain_odd.py (hunk 3 above)
Comment (andiwand) [reply Examples/Scripts/Python/full_chain_odd.py:44]: as you like. I think @timadye  wanted to keep them as close as possible

File: CI/physmon/physmon.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (paulgessinger) [reply CI/physmon/physmon.py]: Refactored.

File: CI/physmon/physmon.py [hunk 12]
```diff
@@ -62,198 +62,80 @@
...
+    ### CKF track finding variations
+
+    for truthSmearedSeeded, truthEstimatedSeeded, label in [
+        (True, False, "truth_smeared"),  # if first is true, second is ignored
+        (False, True, "truth_estimated"),
+        (False, False, "seeded"),
+        (False, False, "orthogonal"),
+    ]:
+        # TODO There seems to be a difference to the reference files when using
+        # multithreading ActsAnalysisResidualsAndPulls
```
[the remark this answers is not in the document]
Comment (paulgessinger) [reply CI/physmon/physmon.py]: Should be gone.

File: Core/include/Acts/Propagator/detail/PointwiseMaterialInteraction.hpp [hunk 13]
```diff
@@ -139,6 +140,10 @@ struct PointwiseMaterialInteraction {
             static_cast<std::underlying_type_t<NavigationDirection>>(nav));
     // put particle at rest if energy loss is too large
     nextP = (mass < nextE) ? std::sqrt(nextE * nextE - mass * mass) : 0;
+    // minimum momentum below which we will not push particles via material
+    // update
+    static constexpr double minP = 50 * Acts::UnitConstants::MeV;
```
Comment (andiwand) [inline Core/include/Acts/Propagator/detail/PointwiseMaterialInteraction.hpp:145]: should we make this configurable? not necessarily in this pr but we can also follow up with a ticket

File: Tests/UnitTests/Core/EventData/MultiTrajectoryComplianceTests.cpp [hunk 14]
```diff
@@ -0,0 +1,88 @@
+// This file is part of the Acts project.
```
Comment (andiwand) [inline Tests/UnitTests/Core/EventData/MultiTrajectoryComplianceTests.cpp:1]: I cant make the connection here. is this file from a different branch?

**Review by andiwand — APPROVED** [review]

File: Core/include/Acts/Propagator/detail/PointwiseMaterialInteraction.hpp (hunk 13 above)
Comment (paulgessinger) [reply Core/include/Acts/Propagator/detail/PointwiseMaterialInteraction.hpp:145]: In the end it's an "epsilon". We can also just make it smaller.

File: Tests/UnitTests/Core/EventData/MultiTrajectoryComplianceTests.cpp (hunk 14 above)
Comment (paulgessinger) [reply Tests/UnitTests/Core/EventData/MultiTrajectoryComplianceTests.cpp:1]: Oh yeah sorry.

File: Core/include/Acts/Propagator/detail/PointwiseMaterialInteraction.hpp (hunk 13 above)
Comment (paulgessinger) [reply Core/include/Acts/Propagator/detail/PointwiseMaterialInteraction.hpp]: Decreased to 10 MeV.

Comment (andiwand) [reply Core/include/Acts/Propagator/detail/PointwiseMaterialInteraction.hpp]: but it relates to if we want to stop particles or not right? so to me it looks like we are slightly changing the physics here. also the `ParticlesStopped` aborter will not fire anymore I guess? https://github.com/acts-project/acts/blob/main/Core/include/Acts/Propagator/StandardAborters.hpp#L206

Comment (andiwand) [reply Core/include/Acts/Propagator/detail/PointwiseMaterialInteraction.hpp]: @paulgessinger one followup question

Comment (paulgessinger) [reply Core/include/Acts/Propagator/detail/PointwiseMaterialInteraction.hpp]: The thing is, if this if not >0, the Stepper update has an floating-point exception and we can't really proceed. So Yeah, this is not ideal, but I don't see another way, other than hard terminating the propagation right here.

**Review by andiwand — APPROVED** [review]

**Pull request merged**